src/day09.rs

```diff
@@ -1,30 +1,364 @@
+        self.marble_ring.get_mut(&two_over).expect("Two over marble was not in map").left_marble = marble_value;
+        
+        self.current_marble = marble_value;
+    }
+
+    fn pop_seventh_and_set_sixth(&mut self) -> i32 {
+        let sixth_left = self.get_nth_left_of_current(6);
+        let seventh_left = self.marble_ring.get(&sixth_left).expect("7th left marble was not in map").left_marble;
+        let eigth_left = self.marble_ring.get(&seventh_left).expect("8th left marble was not in map").left_marble;
+
+        self.marble_ring.remove(&seventh_left);
+
+        self.marble_ring.get_mut(&sixth_left).expect("Sixth left was not in map").left_marble = eigth_left;
+        self.marble_ring.get_mut(&eigth_left).expect("Eigth left marble was not in map").right_marble = sixth_left;
+
+        self.current_marble = sixth_left;
+        seventh_left
+    }
+}
+
+impl Display for HashMarbles {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut result = "".to_string();
+        let first = self.marble_ring.iter().nth(0).unwrap();
+        let mut current = first;
+        loop {
+            if current.0 == first.0 {
+                break
+            } else if *current.0 == self.current_marble {
+                result += format!(" ({})", current.0).as_str();
+            } else {
+                result += format!(" {}", current.0).as_str();
+            }
+            current = (&current.1.right_marble, self.marble_ring.get(&current.1.right_marble).unwrap());
+        }
+        write!(f, "{}", result)
+    }
+}
+struct LListMarbles<'a> {
+    // list: LinkedList<i32>,
+    cursor: CursorMut<'a, i32>,
+}
+
+impl<'a> LListMarbles<'a> {
+    fn new_from_list(list: &'a mut LinkedList<i32>) -> LListMarbles<'a> {
+        list.push_back(0);
+        LListMarbles { cursor:list.cursor_back_mut() }
+    }
+
+    fn move_n_anti_clockwise(&mut self, n: usize) {
+        for _ in 0..n {
+            self.cursor.move_prev();
+            if self.cursor.current().is_none() {
+                self.cursor.move_prev();
+            }
+        }
+    }
+
+}
+
+impl<'a> MarbleCollection for LListMarbles<'a> {
+    fn move_clockwise_and_place_new(&mut self, marble_value: i32) {
+        self.cursor.move_next();
+        if self.cursor.current().is_none() {
+            self.cursor.move_next();
+        }
+        self.cursor.insert_after(marble_value);
+        self.cursor.move_next();
+    }
+
+    fn pop_seventh_and_set_sixth(&mut self) -> i32 {
+        self.move_n_anti_clockwise(7);
+        self.cursor.remove_current().expect("Cursor removed None..!")
+    }
+}
+
+impl <'a> Display for LListMarbles<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // I think it's impossible to display this given that cursor is mutable
+        // and that's the only thing I'm storing in the struct
+        todo!()
```

Comment:
You can call `as_cursor` which will give you a read-only cursor and then iterate through that.